fsspec/implementations/cache_mapper.py

```diff
@@ -30,14 +30,38 @@ def __hash__(self) -> int:
 
 
 class BasenameCacheMapper(AbstractCacheMapper):
-    """Cache mapper that uses the basename of the remote URL.
+    """Cache mapper that uses the basename of the remote URL and a fixed number
+    of directory levels above this.
 
-    Different paths with the same basename will therefore have the same cached
-    basename.
+    The default is zero directory levels, meaning different paths with the same
+    basename will have the same cached basename.
     """
 
+    def __init__(self, directory_levels: int = 0):
+        if directory_levels < 0:
+            raise ValueError(
+                "BasenameCacheMapper requires zero or positive directory_levels"
+            )
+        self.directory_levels = directory_levels
+
+        # Separator for directories when encoded as strings.
+        self._separator = "_^_"
+
     def __call__(self, path: str) -> str:
-        return os.path.basename(path)
+        dirname, basename = os.path.split(path)
+
+        if self.directory_levels > 0:
+            dirs = dirname.split(os.sep)[-self.directory_levels :]
+            dirs.append(basename)
+            basename = self._separator.join(dirs)
+
+        return basename
+
+    def __eq__(self, other: Any) -> bool:
+        return super().__eq__(other) and self.directory_levels == other.directory_levels
+
+    def __hash__(self) -> int:
+        return super().__hash__() ^ hash(self.directory_levels)
```

Comment:
```suggestion
```

Reply:
I haven't made the suggested change yet. Doing so doesn't break any old tests, only the new one I added to explicitly check what I added.

I only added the `__eq__` and `__hash__` functionality because of the same functions on `CachingFileSystem` used to depend on the `same_names` attribute so now depend on the `_mapper` attribute, and for equivalent behaviour `BasenameCacheMapper(1)` must always be equal to any other `BasenameCacheMapper(1)` and different to any `BasenameCacheMapper(2)`. But if we don't need this equality/inequality then it can all go. But then I would ask if this is consistent with the existence of `CachingFileSystem.__eq__`, i.e. what is the purpose of the latter?

Reply:
I think al the eq/hash should be dropped if they are only used for tests or not at all. For the former, we can always look at the value of the attributes, after all; plus every filesystem instance has a `._fs_token_` attribute too.